Add feature for custom formatting the description text

sebastianwahn opened the pull request:
With this feature, it's possible to add a custom formatting to the description text which is submitted to the toggl website.

The current description format looks like:
"TaskId Description" if a task-id and a description is provided (JIRA).

With this feature, a user can set his own formatting. 
The formatting can be activated in the options. Afterwards a user can set the formatting with the currently provided tokens ($$taskId$$ and $$description$$).

A customized format can look like this now:
"$$taskId$ - $$description$$ - test"

![erweiterungen](https://cloud.githubusercontent.com/assets/6577779/15450606/3916d484-1fa0-11e6-83f7-2deb6c33ca67.png)

This will be replaced, so the final description will look like (assuming the taskID is "ABC-123" and description is "Some fancy description"):
"ABC-123 - Some fancy description - test"

![timer_ _toggl](https://cloud.githubusercontent.com/assets/6577779/15450618/c2c70a28-1fa0-11e6-98e6-4d8ccabfa4d0.png)

Currently only the jira-plugin is adjusted to support this feature, since some other sites do not have task-id's  (trello).
To support the formatting for other sites, the member "taskId" must be set on the object passed to createTimerLink()
![banners_and_alerts_und_common_js_-_toggl-button_-____documents_workspace_toggl-button_](https://cloud.githubusercontent.com/assets/6577779/15450639/229f2200-1fa1-11e6-9ce1-57fd10bbc330.png)

tupaschoal commented:
This looks a lot like what I requested at #726 , any plans of allowing this or accepting the pull request?

sebastianwahn commented:
@tupaschoal I'm still waiting for feedback on this PR (except the merge conflicts). Don't know what why a simple 'ok' or 'this is crap' takes that long.

tupaschoal commented:
That's sad @sebastianwahn ):

Can you tell us anything @IndrekV?

IndrekV commented:
Hi guys. Sorry for the lack of feedback. This functionality is based on few tools and we are hesitating to add this as few tools in the list of 90 is a small number. We have kept it open as the idea is great and if there is a way to make more tools benefit from this, it would be a good addition to the extension.

sebastianwahn commented:
@IndrekV i agree that its unfortunate to support only few tools. On the other hand, i don't have accounts for all the tools you are supporting to test the functionality everywhere.
I could add support for some more tools of the list, but i don't think i can implement all.
Maybe, in this case i could add an how-to to add support for more tools?

tupaschoal commented:
@sebastianwahn how straightforward is the addition of new tools? I would definitely be willing to code some to have this feature.

sebastianwahn commented:
@tupaschoal In a nutshell, you only have to modify the render function of the button for a tool. For Jira i modified it in [src/scripts/content/jira.js](https://github.com/toggl/toggl-button/pull/556/files#diff-ac9c1e5274e0954a57c4e4956eb2c20cR20)

tupaschoal commented:
@IndrekV I understand your concern, but there are currently 106 services listed, if I'm not mistaken, and many of them require a subscription. What if you had a branch for me and @sebastianwahn to work on to support more services (the public ones), 

What percentage of the public services would we have to support before this is pushed into the main app?

tupaschoal commented:
?

syko commented:
I'm terribly sorry for the long delay.

We've been reluctant to move forward with this because we're not sure if this setting is too advanced for the average user or if it's beneficial enough for others.

I took a quick look at the other integrations. These are the ones that contain some sort of `id` in the Time Entry description:

- Jira
- Backlog
- Bitbucket
- Desk
- EProject
- Eventum
- Fogbugz
- Freshdesk
- Gemini
- Github
- Gitlab
- Gogs
- Overv
- Phabricator
- PivotalTracker
- Proworkflow
- Redmine
- Smartboard
- Taiga
- TargetProcess
- Teamleader
- Testrail
- Trac
- Unfuddle
- Waffle
- Youtrack
- Zendesk
- Zoho

Additionally, these integrations have different multi-part descriptions:

- Toodledo: `description folderName`
- Trello: `description task`
- Wunderlist: `description task`
- Workfront: `description url`

In order to merge this, we'd like to have support for custom formatting for at least Zendesk, PivotalTracker, Taiga, Gitlab and Github. We can make an issue for implementing the rest later.

Another concern is that if I enable custom formatting `$$taskid$$ - $$description$$` and I use Jira and Todoist, I don't want my Todoist Time Entries to come out as ` - Some description` because of `$$taskid$$` being incompatible with Todoist and thus always empty. We should also strip away so-to-say "dangling separators" then.

Are you guys interested in taking all this on?

Btw I think handlebars-style tokens would look a bit nicer: `{{description}}`.

tupaschoal commented:
Hello @syko,

I'm willing to help if we just concentrate on breaking ground on the feature, that is, working on the 5 you set as minimum to merge this. The others may be doable as time goes on, but I don't think we should wait for all of them. If @sebastianwahn can kindly see what the merge problem is and provide the how-to, I can probably manage adding the other services.

As for the dangling separators, it should be easy to handle. The better option could even be adding a custom formatting tied to a specific service, but we can think this as an enhancement.

What do you say?

sebastianwahn commented:
@syko Thanks for your reply regarding my pull-request. I think implementing your mentioned services shouldn't be a problem.  I have accounts for the following services:
* bitbucket
* trello
* gitlab
* github

For these, i can implement and test the feature. Does a test account exist for Zendesk, PivotalTracker and Taiga? Or are they free of charge? Was not able to check this yet.

Regading the handlebars-style: yes, it looks nicer, didn't thought of it in the first place :)
A fallback if no task-id is given also makes sense.

@tupaschoal I try to handle the merge conflict tomorrow. I also can implement the handlebars support and the fallback if no task-id is given.
For the implementation of the services, we can split up the work, i just need to take a look at the code again.

Sincerely,
Sebastian

tupaschoal commented:
Sounds great @sebastianwahn . Of all the services listed, I have accounts in basically the same ones as you. I've broken up the list made by @syko below to ease ours, or other contributors :), future job:

## Minimum Requirement
- Github ```Free```
- Gitlab ```Free```
- PivotalTracker ```Paid w/ Free Trial```
- Taiga ```Free```
- Zendesk ```Paid w/ Free Trial```

## Good to Have
- Jira ```Paid w/ Free Trial```
- Backlog ```Paid w/ Free Trial```
- Bitbucket ```Free```
- Desk ```Paid w/ Free Trial```
- EProject ```Paid w/ Free Trial```
- Eventum ```Free```
- Fogbugz ```Paid w/ Free Trial```
- Freshdesk ```Paid w/ Free Trial```
- Gemini ```Paid w/ Free Trial```
- Gogs ```Free```
- Overv ```Free```
- Phabricator ```Free```
- Proworkflow ```Paid w/ Free Trial```
- Redmine ```Free```
- Smartboard ```Free```
- TargetProcess ```Paid w/ Free Trial```
- Teamleader ```Paid w/ Free Trial```
- Testrail ```Paid w/ Free Trial```
- Trac ```Free```
- Unfuddle ```Free```
- Waffle ```Free```
- Youtrack ```Free```
- Zoho ```Paid w/ Free Trial```

sebastianwahn commented:
@tupaschoal Merge Conflicts are fixed now.

syko commented:
Sounds good. 

> As for the dangling separators, it should be easy to handle. The better option could even be adding a custom formatting tied to a specific service, but we can think this as an enhancement.
>
>What do you say?

I wouldn't want to add that since it makes configuring the extension overly complex.

**To elaborate on the need to trim separators:**

Different tools could have different tokens. For example Trello has an extra "task" but only when tracking time on a checklist item. I don't want to add per-service formatting so this means that one custom format string needs to work for all services while behaving well if a service doesn't support a given token.

For example if I'm using Jira and Trello at the same time and want to make sure of Jira's ids and Trello's tasks I should be able to specify a format string like `{{taskId}}: {{description}} - {{task}}` and it would produce Time Entries like `123ABC: some description` for Jira, `some description - some checklist item` for trello and `some description` for services that don't support either of these tokens.

With this kind of well-behaved fallback it is not even necessary to turn the custom formatting "on" or "off" as it will always work. (we should have that checkbox in the UI for UX purposes but the checkbox can simply apply a default format instead of disabling the feature)

Also, perhaps a `{{service}}` token which would be replaced by the name of the tool (Jira/Asana) would also be useful to have? And maybe rename `{{taskId}}` to just `{{id}}` so as not to confuse it with Toggl's tasks or Trello's tasks and such.

tupaschoal commented:
I agree with your suggestions @syko, how do we proceed to work on this as it is not in the main repo and I can't push to this PR?

Thank you for handling the merge @sebastianwahn !

leafi commented:
@tupaschoal Oof. Good point.. that's a hairy permissions issue.

Either @sebastianwahn could open his clone of toggl-button to you & I guess us too, or we could clone toggl-button and have e.g. github.com/toggl/toggl-button-custom-formatting, add you and Sebastian as collaborators there, and set up a new PR going from: toggl-button-custom-formatting to: toggl-button.

I guess the latter would also deal with the inevitable problem of us needing to rewrite the commit log come merge time.

@syko : thoughts?

syko commented:
I don't think github lets you fork your own repo (within the same organization) so it's easiest if @sebastianwahn & @tupaschoal handle that.

I could fork this fork under my own account and give access but I don't see a point there. As long as you keep the commit history somewhat decent (squashed, yet atomic as much as possible) we should be fine.

tupaschoal commented:
@sebastianwahn could you please add me as a contributor in your repo? I think this would be the easier way, pushing to yours and then making sure commits there are being added to this PR.

I have worked on all the

tupaschoal commented:
I've done some more changes:
```
533589eAdd custom format support to Trac
4bfaab9Add custom format support to Redmine
8bfbef9Add custom format support to Phabricator
4f00a2bAdd custom format support to Gogs
bc9ee53Add custom format support to BitBucket
9c3c5e9Add how-to use custom format and supported services to README
```
Here is the updated checklist:
## Minimum Requirement
- [x] Github ```Free```
- [x] Gitlab ```Free```
- [x] PivotalTracker ```Paid w/ Free Trial```
- [x] Taiga ```Free```
- [x] Zendesk ```Paid w/ Free Trial```

## Good to Have
- [x] Jira ```Paid w/ Free Trial```
- [ ] Backlog ```Paid w/ Free Trial```
- [x] Bitbucket ```Free```
- [ ] Desk ```Paid w/ Free Trial```
- [ ] EProject ```Paid w/ Free Trial```
- [ ] Eventum ```Free```
- [ ] Fogbugz ```Paid w/ Free Trial```
- [ ] Freshdesk ```Paid w/ Free Trial```
- [ ] Gemini ```Paid w/ Free Trial```
- [x] Gogs ```Free```
- [ ] Overv ```Free```
- [x] Phabricator ```Free```
- [ ] Proworkflow ```Paid w/ Free Trial```
- [x] Redmine ```Free```
- [ ] Smartboard ```Free```
- [ ] TargetProcess ```Paid w/ Free Trial```
- [ ] Teamleader ```Paid w/ Free Trial```
- [ ] Testrail ```Paid w/ Free Trial```
- [x] Trac ```Free```
- [ ] Unfuddle ```Free```
- [ ] Waffle ```Free```
- [ ] Youtrack ```Free```
- [ ] Zoho ```Paid w/ Free Trial```

sebastianwahn commented:
@tupaschoal i will add you as a contributor now. Unfortunately, you also implemented some of the things i already implemented (handlebars, id, ..) so this will lead to a merge conflict.

When i tested the jira integration with the {{service}} token, all worked fine on my site. But currently, i am working on a feature branch.

tupaschoal commented:
It is alright @sebastianwahn. I'll separate the non-conflicting commits tomorrow and push to the repo. Would you rather have me pushing to the feature/customFormat branch or directly to the master?

sebastianwahn commented:
@tupaschoal please create a separate feature branch, then i can review your code. Additionally,  please create a pull request from your feature branch to my master, if this is possible within the same repository.

tupaschoal commented:
As you wish @sebastianwahn. I created a new branch and was able to create a pull request as well. I'm not being able to test the formatting, so every commit there has been tested with the custom format on.

I've migrated most of the diffs I already had :)

tupaschoal commented:
Hello @syko and @leafi, is this enough for the PR to be accepted? :)

syko commented:
Hey. Wonderful, that's quite a handful of services covered :+1:

Sry for the delay once more, but I'll get to reviewing the PR as soon as possible!

tupaschoal commented:
Where do we stand here? :)

tupaschoal commented on src/scripts/content/gitlab.js:
```diff
@@ -25,18 +27,20 @@ togglbutton.render('.issue-details .detail-page-description:not(.toggl)', {obser
   if (numElem !== null) {
-    description = "MR" + numElem.textContent.split(" ").pop().trim().replace("!", "") + "::" + description;
+    id = "MR" + numElem.textContent.split(" ").pop().trim().replace("!", "");
+    description = "::" + description;
```
Agreed!

tupaschoal commented on src/scripts/settings.js:
```diff
@@ -712,6 +717,13 @@ document.addEventListener('DOMContentLoaded', function (e) {
     });
 
+    Settings.$useCustomFormat.addEventListener('click', function (e) {
+        Settings.toggleSetting(e.target, (localStorage.getItem("useCustomFormat") !== "true"), "toggle-use-custom-format");
```
I think it makes the file inconsistent. @sebastianwahn just probably followed the common practice, which you can see from lines 604-635. I changed it but reverted after I saw this.

tupaschoal commented on src/scripts/settings.js:
```diff
@@ -712,6 +717,13 @@ document.addEventListener('DOMContentLoaded', function (e) {
+        Settings.toggleSetting(e.target, (localStorage.getItem("useCustomFormat") !== "true"), "toggle-use-custom-format");
+    });
+    document.querySelector("#custom-format").addEventListener('blur', function (e) {
+        Settings.saveSetting((document.querySelector('#custom-format').value), "toggle-custom-format");
```
Done.

tupaschoal commented on src/scripts/content/gemini.js:
```diff
@@ -3,18 +3,20 @@
   if (numElem !== null) {
-    description = numElem.textContent + " " + description;
+    id = numElem.textContent.split(" ")[0];
+    description = description.split(" ").splice(1);
```
I tested all of them and they worked, if they didn't I'd not push. I could try to fix it as you suggested, but my account on Gemini has expired, so I wouldn't be able to test :(

tupaschoal commented:
I created a PR on @sebastianwahn's repo that fixes 2 of the 4 issues (and have commented on the ones I didn't above. We also need to solve the bitbucket conflict, which seems to be straightforward enough.

sebastianwahn commented:
@syko 

> See the comments about code. Additionally, how about that separator trimming I was talking about before? #556 (comment)
Right now I'm gonna get : some description if I put down {{id}}: {{description}} as my custom format and try tracking from todoist

I thought again about this problem:

What happens if there is not only a separator, but also some static text which needs to be there `{{id}} - my kind of awesome text - {{description}} - {{service}}`

In this case, if the `{{id}}` is null and i trim everything from `{{id}}` to the start of `{{description}}` the text would be also gone.
Currently, i don't see a solution to just trim the separators, since those are defined by the user and not restricted to just `-` or `:`. Therefore I have to either guess the separator, but what if there is none? For example with ` ` (blank) as separator. Otherwise, I have to delete everything from one token to the next token (`{{id}}` is not set -> delete everything up to `{{description}}`) resulting in missing static text within the formatting. Problem with the deletion from one token to the next is also I don't know if the user defined `{{id}}` as a first token or as the last (because he can).

syko commented:
Yeah I thought about that but I don't see a good use case for including custom text in the description format. I'd much rather not have that as a feature than end up with untrimmed separators in some services.

> Problem with the deletion from one token to the next is also I don't know if the user defined {{id}} as a first token or as the last (because he can).

Not sure I understand this bit. You could first replace all unsupported tokens with `''`, trim the string of non-token characters using regex and then replace the supported tokens with values wherever they are in the string.

wareczek commented:
Please add token `{{url}}`

bryanph commented:
> @syko
> 
> > See the comments about code. Additionally, how about that separator trimming I was talking about before? #556 (comment)
> > Right now I'm gonna get : some description if I put down {{id}}: {{description}} as my custom format and try tracking from todoist
> 
> I thought again about this problem:
> 
> What happens if there is not only a separator, but also some static text which needs to be there `{{id}} - my kind of awesome text - {{description}} - {{service}}`
> 
> In this case, if the `{{id}}` is null and i trim everything from `{{id}}` to the start of `{{description}}` the text would be also gone.
> Currently, i don't see a solution to just trim the separators, since those are defined by the user and not restricted to just `-` or `:`. Therefore I have to either guess the separator, but what if there is none? For example with (blank) as separator.
> Otherwise, I have to delete everything from one token to the next token (`{{id}}` is not set -> delete everything up to `{{description}}`) resulting in missing static text within the formatting. Problem with the deletion from one token to the next is also I don't know if the user defined `{{id}}` as a first token or as the last (because he can).

I think a per-service formatting string would actually be nicer because we wouldn't have to do this sub-optimal text manipulation. We could detect which service the current page is running and then show the associated "settings" page.

I created an issue for this PR (see #1187) and I'll run it by the front-end team again to see if we can implement this soon. We're sorry for the long back and forth. Opening an issue will hopefully make us pick it up soon! 

For now I will close this PR, feel free to reopen if you wish to continue work once the discussion on the issue has resolved. We'll make sure to give you credit when we do merge this change.

tupaschoal commented:
I like this idea, if it gets out of paper, I'm on it.
